![mumbai 26/11](https://akm-img-a-in.tosshub.com/indiatoday/images/story/201904/Mumbai_terror_attack_Reuters.jpeg?CVg_HP.Ovcc.2zQ35Ohoe0GRzCm5YInJ)

# What is terrorism?


Terrorism is definitely a some kind of violence. To understand terrorism we need a consistent and clear definition of what terrorism is and how it is different from any other form of violence. This is not straightforward.

Terrorism is defined in the Oxford Dictionary as **“the unlawful use of violence and intimidation, especially against civilians, in the pursuit of political aims.”**


As per the definition by the United Nations, **“any criminal acts intended or calculated to provoke a state of terror in the general public, a group of persons or particular persons for a particular purpose are in any circumstance unjustifiable, whatever the considerations of a political, philosophical, ideological, racial, ethnic, religious or any other nature that may be invoked to justify them.”**

Yes!  These definitions try to accommodate almost everything and very difficult to understand. Must be written by a bunch of lawyers, that’s why it looks very complicated. 

These definitions do not provide any clear understanding and very subjective. There is no internationally recognized legal definition of terrorism. According to one country, an act is a terrorist activity and according to others, it may not. So there is no general consensus among countries.



Let’s try to take out common between them,

To be considered an act of terrorism, an action must be violent or threaten 


1. **To be considered an act of terrorism, an action must be violent or threaten violence. As such, political dissent, activism, and nonviolent resistance do not constitute terrorism.**
2. **The inclusion of damage to private and public property.**
3. **Action must also be carried out for political, economic, religious, or social purposes to count as terrorism.**
4. **Designed to have far-reaching psychological repercussions beyond the immediate victim or target.**
5. **Targetting noncombatant, neutral, or randomly chosen people.**
6. **The action must be perpetrated by a subnational group or non-state entity.** 

we have data of various terrorist activities which happened between 1970 and 2017. Lets explore it.

In [ ]:
# importing libraries
import numpy as np
import pandas as pd
import plotly.express as px

# Data Reading

In [ ]:
# reading csv

terrorism = pd.read_csv("/content/drive/MyDrive/Sparks Foundation task/Exploratory Data Analysis - Terrorism Task-4/globalterrorismdb_0718dist.csv",
                        encoding="ISO-8859-1",low_memory=False)

print(terrorism.shape)
terrorism.head()

(181691, 135)


eventid  iyear  imonth  iday  ... INT_IDEO  INT_MISC INT_ANY  related
0  197000000001   1970       7     2  ...        0         0       0      NaN
1  197000000002   1970       0     0  ...        1         1       1      NaN
2  197001000001   1970       1     0  ...       -9         1       1      NaN
3  197001000002   1970       1     0  ...       -9         1       1      NaN
4  197001000003   1970       1     0  ...       -9         1       1      NaN

[5 rows x 135 columns]

**we have data of 181691 horrific events happened between 1970 and 2017 around the world.**

# Terrorism deaths globally

In [ ]:
# lets find out deaths in recent past
labels = ['1970-1979','1980-1989','1990-1999','2000-2009','2010-2017']
bins = [1969,1979,1989,1999,2009,2019]
terrorism['decade'] = pd.cut(terrorism['iyear'],bins=bins,labels=labels)

In [ ]:
# lets find out how many human lives terrorism took 

human_death = terrorism.groupby(['iyear'])['nkill'].sum()
human_death = human_death.reset_index()

fig = px.line(human_death, x='iyear', y='nkill', title='Deaths from terrorism, 1970 to 2017', template="plotly_dark", 
              labels=dict(iyear="YEAR", nkill="No. of Kills"))

fig.update_xaxes(rangeslider_visible=True)
fig.update_traces(mode="markers+lines")

fig.show()

In [ ]:
human_death = terrorism.groupby(['iyear', 'country_txt','region_txt'])['nkill'].sum()
human_death = human_death.reset_index()

fig = px.choropleth(human_death, locations="country_txt",locationmode="country names", color="nkill", 
                    animation_frame="iyear", range_color=[0,700],
                    labels=dict(iyear="YEAR", nkill="No. of Kills", country_txt="Country",region_txt='Region'),
                    color_continuous_scale="reds",
                    template="plotly_dark", 
                    hover_data=['region_txt'],
                    title='Deaths from terrorism 1970-2017')


fig.update_layout(height=600,)
fig.show()

In [ ]:
human_death = terrorism.groupby(['decade'])['nkill'].sum()
human_death = human_death.reset_index()

fig = px.pie(human_death, values='nkill', 
             names='decade', title='Deaths from terrorism, 1970 to 2017', 
             labels=dict(decade="Decade", nkill="No. of Kills"),
             template="plotly_dark")
fig.show()

In 2017, an estimated **26,445** people died from terrorism globally. From 1970 till 2017 **411868** peolpe died, **48 %** of them died in previous decade(2010-2017). 

However, there can be significant year-to-year variability. In **2014** approximately **44.49k people died**,which is hihest till now. 

sharp increase was largely due to groups like **Islamic State in Iraq** and **Boko Haram in Nigeria** bold text.The deadliest attack was in the Iraqi city of Mosul, when Islamic State killed 670 Shia prisoners.

"Unprecedented" seizure of territory in Iraq and Syria by ISIS, the flow of foreign terrorists joining the group and "lone" offender extremists in the West were major trends in global terrorism in 2014, according to the US state department report.

Al-Qaeda was largely fragmented in 2014, but failed governments in countries like Yemen, Syria, Libya, Nigeria and Iraq enabled extremist radicalism.

**Terrorist groups used more aggressive tactics in 2014 than years before, such as beheadings and crucifixions.**

# Global Distribution of Terrorism

As we observe above in 2017 approximately 26k people died due to terrorism. lets find out where in the world did terrorists kill most people?

In [ ]:
human_death = terrorism.groupby(['iyear', 'country_txt','region_txt'])['nkill'].sum()
human_death = human_death.reset_index()
human_death_dist = human_death.groupby(['iyear','region_txt'])['nkill'].sum()
human_death_dist = human_death_dist.reset_index()
human_death_dist.sort_values(["iyear","nkill"],ascending=[True,False],inplace=True)

fig = px.bar(human_death_dist, x='nkill', y='region_txt',animation_frame="iyear",orientation='h',template="plotly_dark",color='region_txt',
             labels=dict(iyear="YEAR", nkill="No. of Kills",region_txt='Region',) ,
             title='Global Distribution of Terrorism').for_each_trace(lambda t: t.update(name=t.name.replace("Region=","")))


fig.show()


In [ ]:
fig = px.line(human_death_dist, x='iyear', y='nkill', title='Global Distribution of Terrorism', template="plotly_dark",color='region_txt',
              hover_data=['nkill','region_txt'],
              labels=dict(iyear="YEAR", nkill="No. of Kills",region_txt='Region')).for_each_trace(lambda t: t.update(name=t.name.replace("Region=","")))

fig.update_xaxes(rangeslider_visible=True)


fig.show()

- **Out of 26k** people who died in 2017, **10.8k** are from the **Middle East and North Africa.**
- Numbers in 2014 are more disturbing. **The Middle East and North Africa, South Asia, Sub-Saharan Africa regions were heavily affected.**
- **Australia and Oceania** have seen a **very low number of activities** over the **last 5 decades**. 
- Overall we can see an **increasing trend of terrorism activities**. 


-----------------------------------------------------------------


- The situation in the Middle East and North Africa is the most critical one, we can see **two peaks one in 2007 and another one in 2011.**
- In 2007 by the US president announced **"The New Way Forward"** policy and deployed more than 20k soldiers in Iraq.
- **The Iraqi insurgency** was an insurgency that began in late 2011 after the end of the Iraq War and the withdrawal of U.S. troops from Iraq.
- **Resulting in violent conflict with the central government, as well as low-level sectarian violence among Iraq's religious groups.** & this power struggle is still going on.



## Terrorism as a share of total deaths

In [ ]:
human_death = terrorism.groupby(['iyear', 'country_txt','region_txt'])['nkill'].sum()
human_death = human_death.reset_index()
human_death_1 = human_death.groupby(['iyear', 'country_txt'])['nkill'].sum()
human_death_percent = human_death_1.groupby(level=0).apply(lambda x:
                                                 100 * x / float(x.sum()))
human_death_percent= human_death_percent.reset_index()

fig = px.choropleth(human_death_percent, locations="country_txt",locationmode="country names", color="nkill", 
                    animation_frame="iyear", range_color=[0,15],
                    labels=dict(iyear="YEAR", nkill="Percentage Kills", country_txt="Country"),
                    color_continuous_scale="reds",
                    template="plotly_dark", 
                    hover_data=['country_txt'],
                    title='Country wise Percentage deaths from terrorism')


fig.update_layout(height=600,)
fig.show()

**Iraq** – the country with the **most fatalities in 2017** – accounted for **60%** of deaths in the Middle East & North Africa. This was **one-in-four** terrorism deaths globally. Combined, Iraq and Syria accounted for nearly **80%** in the region, and **one-in-three** globally.

In US terrorism deaths in most years are very few: typically below **0.01%** of all deaths. This unexpectedly spiked with the **9/11 attacks** – **the world’s most fatal terrorist event of recent times. It claimed 3000 lives.**

As we have observed earlier The Middle East and North Africa, South Asia, Sub-Saharan Africa regions were heavily affected. But not all the countries from these regions were affected.

**In South Asia**, most deaths occurred in **Afghanistan**, with high numbers in **Pakistan** and **India** too. But some countries in the region – such as **Nepal** – had almost none. In the middle east region, **Syria and Iraq** were mostly affected.

In 1999 out of total deaths due to terrorism around the world **13% was alone from India.** India faced tragic events such as the **Kargil war** and the **Indian Airlines Flight 814 highjacking** incident. 



In [ ]:
# deaths from terrorism decade wise analysis
human_death = terrorism.groupby(['decade', 'country_txt','region_txt'])['nkill'].sum()
human_death = human_death.reset_index()

fig = px.choropleth(human_death, locations="country_txt",locationmode="country names", color="nkill", 
                    animation_frame="decade", range_color=[0,700],
                    labels=dict(iyear="YEAR", nkill="No. of Kills", country_txt="Country",region_txt='Region'),
                    color_continuous_scale="reds",
                    template="plotly_dark", 
                    hover_data=['region_txt'],
                    title='Deaths from terrorism decade wise')


fig.update_layout(height=600,)
fig.show()

- In the **1970s** the focus of terrorism was on **North, South America, Europe, and Iran.**
- In the **1980s** the terrorism spread accross the globe, **India and Peru were most affected.**
- In the **1990s** it spread like wildfire in **African countries**. Focus on North America and Europe was fading.
- In the **2000s** it starts to spread again. Hitting top countries like the **US, Russia, India** and for the first time, **Southeast Asian countries** see an unprecedented increase of events.
- In the **2010s** it is focused on **Asia** and a **larger part of Africa**.

-----------------------------------------------------------------------------

**To understand the reasons behind this spread, we need to understand the geopolitical events that happened during the last 5 decades. The majority of the time we can see high levels of internal conflicts were going on within the countries. The cause of those conflicts was religious, political, ideological or ethnic, etc.**

# Hot Zones around the World

In [ ]:
# kill wise
hot_zones = terrorism.groupby(['country_txt','city','latitude','longitude'])['nkill'].sum()
hot_zones = hot_zones.reset_index()
hot_zones.sort_values(by='nkill', ascending=False,inplace=True)
hot_zones = hot_zones.head(100)

fig = px.scatter_geo(hot_zones, lat="latitude", lon="longitude",color="country_txt",
                  color_continuous_scale=px.colors.cyclical.IceFire, size='nkill', range_color=[0,1000],height=550,
                  hover_name="city", template="plotly_dark",
                  
                  labels=dict( nkill="No. of Kills", country_txt="Country",city='City'),
                  title='Hot Zone Cities around the world - Severity').for_each_trace(lambda t: t.update(name=t.name.replace("Country=","")))
fig.show()

**We can observe that Baghdad is a massive hot zone, with 20.8k deaths in the last 5 decades. After that Mosul having 7k fatalities.**

**We can also observe Mogadishu in Somalia, New York City, Belfast in the UK, Karachi, Kabul are hot zones.**

In [ ]:
# number of events wise
hot_zones = terrorism.groupby(['country_txt','city','latitude','longitude'])['eventid'].count()
hot_zones = hot_zones.reset_index()
hot_zones.sort_values(by='eventid', ascending=False,inplace=True)
hot_zones = hot_zones.head(100)

fig = px.scatter_geo(hot_zones, lat="latitude", lon="longitude",color="country_txt",
                  color_continuous_scale=px.colors.cyclical.IceFire, size='eventid', range_color=[0,1000],height=550,
                  hover_name="city", template="plotly_dark",
                  
                  labels=dict( eventid="No. of Terrorism Activities", country_txt="Country",city='City'),
                  title='Hot Zone Cities around the world - No. of Terrorism Activities').for_each_trace(lambda t: t.update(name=t.name.replace("Country=","")))
fig.show()

**We can observe, In the last five decades Baghdad-Iraq, Belfast-UK, Karachi-Pakistan, Lima-Peru, Mogadishu-Somalia, etc. had encountered a very high number of terrorism activities.**


# Most violent terrorist organizations around the world

In [ ]:
terror_groups = terrorism.groupby(['gname'])['nkill'].sum()
terror_groups = terror_groups.reset_index()
terror_groups.sort_values(["nkill"],ascending=[False],inplace=True)
terror_groups=terror_groups.head(11)
terror_groups = terror_groups.iloc[1:,:] # removing unknown sources
terror_groups = terror_groups.reset_index()
terror_groups.sort_values(["nkill"],ascending=[True],inplace=True)

fig = px.bar(terror_groups, x='nkill', y='gname',orientation='h',template="plotly_dark",color='nkill',
             labels=dict(nkill="No. of Kills",gname='Organization',) ,
             title='Most violent terrorist organizations around the world')


fig.show()

**As we can see ISIL is the most violent terrorist organization around the world, It has killed almost 39k people till now, and we are not even counting the injured people. Horrific**

**Followed by Taliban, Boko haram, SL, LTTE. together they have killed over 1 lakh people till now.**


# Attack methods preferred by terrorist 

In [ ]:
terror_attack = terrorism.groupby(['decade','region_txt','attacktype1_txt'])['eventid'].count()
terror_attack = terror_attack.reset_index()
terror_attack.rename(columns={'eventid': 'count'}, inplace=True)
terror_attack.sort_values(["decade","count"],ascending=[True,True],inplace=True)

fig = px.bar(terror_attack, x='count', y='attacktype1_txt',animation_frame="decade",orientation='h',template="plotly_dark",color='region_txt',
             labels=dict(decade="Decade", attacktype1_txt="Type of Attack",count = "Count",region_txt='Region') ,
             title='Type of Attack Preferred by Terrorist').for_each_trace(lambda t: t.update(name=t.name.replace("Region=","")))


fig.show()

- In the 70s bombing/explosion, assassinations were the preferred way of terrorism.

- In the 80s along with bombing and assassination, armed assault activities became very frequent. Maybe during this period terrorist organizations somehow got access to arms and ammunition. The frequency of hostage-taking and attacks on facilities increased a little bit.

- In the 90 there was a sharp increase in the above activities. Also, the number of hostage-taking incidents also grew. some incidents of hijacking also took place in this period.

- In the 2000s bombing and armed assault were the preferred to carry out terrorist activities.

- In the 2010s almost all kinds of mentioned activities increased sharply. Especially hostage-taking and attacks on facilities. 

**The main objective of terrorist organizations is to spread terror, fear. In recent times they are trying to use all kinds of ways to achieve it.  By taking hostages mostly foreign country citizens and attacking facilities such as foreign embassies they are directly challenging the particular countries.** 


# Who do they target?

In [ ]:
terror_target = terrorism.groupby(['decade','targtype1_txt'])['eventid'].count()
terror_target = terror_target.reset_index()
terror_target.rename(columns={'eventid': 'count'}, inplace=True)
terror_target.sort_values(["decade","count"],ascending=[True,False],inplace=True)

fig = px.bar(terror_target, x='count', y='targtype1_txt',animation_frame="decade",orientation='h',template="plotly_dark",color='targtype1_txt',
             labels=dict(decade="Decade", targtype1_txt="Target",count = "Count",region_txt='Region') ,
             title='Target of Terrorism').for_each_trace(lambda t: t.update(name=t.name.replace("Target=","")))

fig.update_layout(height=650,)
fig.show()

- It looks like **business** is their main Target. Over the period of 5 decades, the number of attacks on businesses and private citizens was quite high.

- From the 80s onwards their engagement with the **military, police, transportation, and utilities** was incersaed gradually. 

- In the recent past number of direct engagements with the government was more frequent. It looks like they have become more stronger and confident.

- In the recent past, they were targeting **educational institutions** more frequently. 
------------------------------------------------------------------------------

**The overall growth of any country depends upon its performance in social, political, and economic sectors. Only in a stable environment, a country can grow.**

**By targeting business, utilities, and transport they are creating instability in the economic sector. By targeting the military, police, they are directly challenging political stability. By Targetting educational institutions they are spreading fear among the future generation.** 



-----------------------------------------------------------------------------------

-----------------------------------------------------------------------------------

# SOUTH ASIA

**let's explore how terrorism is impacting the South Asia.**

In [ ]:
south_asia = terrorism[terrorism['region_txt']=="South Asia"]
print(south_asia.shape)
south_asia.head()

(44974, 136)


eventid  iyear  imonth  iday  ... INT_MISC  INT_ANY related     decade
585   197011010001   1970      11     1  ...        1        1     NaN  1970-1979
1186  197202220004   1972       2    22  ...        1        1     NaN  1970-1979
1863  197305010002   1973       5     1  ...        1        1     NaN  1970-1979
2217  197402020001   1974       2     2  ...        1        1     NaN  1970-1979
2705  197412090001   1974      12     9  ...        1        1     NaN  1970-1979

[5 rows x 136 columns]

**we have record of 44974 terrorism events happened in South Asia.**

In [ ]:
px.set_mapbox_access_token("pk.eyJ1IjoiazJqMiIsImEiOiJja2dqNHBkeW0wNjByMnJudmU0aDUyZW5zIn0.cS0GcimN3eHU9yupRU_x3g")

fig = px.scatter_mapbox(south_asia, lat="latitude", lon="longitude",color="nkill",
                  color_continuous_scale=px.colors.cyclical.IceFire, size_max=5, zoom=2.5,range_color=[0,10],mapbox_style="light",height=550,
                  hover_name="city",
                  labels=dict(nkill="No. of Kills", targtype1_txt="Target"),
                  title='Death From Terrorism - South Asia')
fig.show()

**On a first look, we can see that Indias North, East, and Nort-East part got clustered by terrorism events. Almost all parts of Afghanistan, Pakistan, Bangladesh, Srilanka are covered by such events.**


**Bhutan and Nepal look good, relatively stable countries.**


In [ ]:
south_asia_trend = south_asia.groupby(['iyear', 'country_txt','region_txt'])['nkill'].sum()
south_asia_trend = south_asia_trend.reset_index()
south_asia_trend = south_asia_trend.groupby(['iyear','country_txt'])['nkill'].sum()
south_asia_trend = south_asia_trend.reset_index()
south_asia_trend.sort_values(["iyear","nkill"],ascending=[True,False],inplace=True)

fig = px.line(south_asia_trend, x='iyear', y='nkill', title='South Asia Distribution of Terrorism', template="plotly_dark",color='country_txt',
              hover_data=['nkill','country_txt'],
              labels=dict(iyear="YEAR", nkill="No. of Kills",country_txt='Country')).for_each_trace(lambda t: t.update(name=t.name.replace("Country=","")))

fig.update_xaxes(rangeslider_visible=True)


fig.show()

**Between mid 80 to mid 90, we can see more frequent activities in south Asia.  Now in the recent past situation in Pakistan and Afghanistan looks serious. India had faced a 2008 terrorist attack, the biggest of all time for india. But after 2010 frequency of such events has been decreased.**

In [ ]:
south_asia_trend = south_asia.groupby(['decade', 'country_txt','region_txt'])['nkill'].sum()
south_asia_trend = south_asia_trend.reset_index()
south_asia_trend = south_asia_trend.groupby(['decade','country_txt'])['nkill'].sum()
south_asia_trend = south_asia_trend.reset_index()
south_asia_trend.sort_values(["decade","nkill"],ascending=[True,False],inplace=True)

fig = px.line(south_asia_trend, x='decade', y='nkill', title='South Asia Distribution of Terrorism', template="plotly_dark",color='country_txt',
              hover_data=['nkill','country_txt'],
              labels=dict(decade="DECADE", nkill="No. of Kills",country_txt='Country')).for_each_trace(lambda t: t.update(name=t.name.replace("Country=","")))

fig.update_xaxes(rangeslider_visible=True)


fig.show()

**In the last decade, there is an unprecedented rise in terrorism activities in Afghanistan and Pakistan. For India, we can observe a decreasing trend.**


In [ ]:
hot_zones_sa = south_asia.groupby(['country_txt','city','latitude','longitude'])['nkill'].sum()
hot_zones_sa = hot_zones_sa.reset_index()
hot_zones_sa.sort_values(by='nkill', ascending=False,inplace=True)
hot_zones_sa = hot_zones_sa.head(100)

fig = px.scatter_geo(hot_zones_sa, lat="latitude", lon="longitude",color="country_txt",
                  color_continuous_scale=px.colors.cyclical.IceFire, size='nkill', range_color=[0,1000],height=550,
                  hover_name="city", template="plotly_dark",
                  
                  labels=dict( nkill="No. of Kills", country_txt="Country",city='City'),
                  title='Hot Zone Cities in South Asia - Severity').for_each_trace(lambda t: t.update(name=t.name.replace("Country=","")))
fig.show()


**In the South Asia region Pakistan, Afghanistan followed by Sri Lanka looks hot zone of terrorism, having very high fatalities.** 

**Karachi, Quetta, Kabul, Kandahar, Colombo, Mumbai were affected by terrorism activities.**


In [ ]:
hot_zones_sa = south_asia.groupby(['country_txt','city','latitude','longitude'])['eventid'].count()
hot_zones_sa = hot_zones_sa.reset_index()
hot_zones_sa.sort_values(by='eventid', ascending=False,inplace=True)
hot_zones_sa = hot_zones_sa.head(100)

fig = px.scatter_geo(hot_zones_sa, lat="latitude", lon="longitude",color="country_txt",
                  color_continuous_scale=px.colors.cyclical.IceFire, size='eventid', range_color=[0,1000],height=550,
                  hover_name="city", template="plotly_dark",
                  
                  labels=dict( eventid="No. of Terrorism Activities", country_txt="Country",city='City'),
                  title='Hot Zone Cities in South Asia - No. of Terrorism Activities').for_each_trace(lambda t: t.update(name=t.name.replace("Country=","")))
fig.show()

**We can observe that in the South Asia region Karachi, kabul, Srinagar, Dhaka, Colombo encountered a very high number of terrorism activities in the last five decades.**

In [ ]:
terror_groups_sa = south_asia.groupby(['gname'])['nkill'].sum()
terror_groups_sa = terror_groups_sa.reset_index()
terror_groups_sa.sort_values(["nkill"],ascending=[False],inplace=True)
terror_groups_sa=terror_groups_sa.head(11)
terror_groups_sa = terror_groups_sa.drop([580], axis=0) # removing unknown sources
terror_groups_sa = terror_groups_sa.reset_index()
terror_groups_sa.sort_values(["nkill"],ascending=[True],inplace=True)

fig = px.bar(terror_groups_sa, x='nkill', y='gname',orientation='h',template="plotly_dark",color='nkill',
             labels=dict(nkill="No. of Kills",gname='Organization',) ,
             title='Most violent terrorist organizations - South Asia ')


fig.show()

**Taliban, LTTE and TTP dominating the Indian Subcontinent.**

**In India, we can see the presence of Sikh extremists during the 70s, LTTE during the 90s to late 2010s. Maoist organizations also spreading terrorism in India.**
